transport/server.cc

```diff
@@ -724,11 +724,36 @@ future<> cql_server::connection::process_request() {
         }
 
         auto fut = get_units(_server._memory_available, mem_estimate);
+        struct request_releaser {
+            timer<lowres_clock> timer;
+            bool timed_out = false;
+        };
+        ::shared_ptr<request_releaser> releaser;
         if (_server._memory_available.waiters()) {
+            releaser = ::make_shared<request_releaser>();
+            releaser->timer.set_callback([this, releaser, mem_estimate] {
+                releaser->timed_out = true;
+                _server._memory_available.signal(mem_estimate);
+                ++_server._stats.requests_shed;
+            });
+            //FIXME: Instead of a hardcoded value, it would be nice to parse the timeout from the message - which may require CQL parsing, i.e. not trivial.
+            //FIXME: Concurrency of this path is 1 and thus releasing 1 request per 5 seconds makes very little sense on its own. It would be better if requests already
```

Comment:
If requests on the same connection have different timeouts, we will end up with just the long timeouts on the list (since the short timeouts will be expired). So IMO this needs to be done on the session level, not the request level.

Reply:
Good idea, then we'll have access to per-session timeouts which would give much more context for free.

Reply:
This idea will be easily applicable after https://github.com/scylladb/scylla/pull/7913 is merged (which will in turn only happen after https://github.com/scylladb/scylla/pull/7867), because then the client state, accessible from this function, will keep its own timeouts. By the time we take the semaphore the query is not yet parsed, so we'd either have to do that in order to know which timeout value to pick (and push the admission semaphore after the parsing stage), or just apply some heuristics - e.g. take max() of all timeouts, which is guaranteed to not fail any query prematurely.